Sources/GRPC/Concurrency/AsyncWriter.swift

```diff
@@ -0,0 +1,257 @@
+  internal func resume() {
+    self.storage.resume()
+  }
+
+  /// Write an `element`.
+  ///
+  /// The call may be suspend if the writer is paused.
+  ///
+  /// Throws: ``AyncWriterError`` if the writer has already been finished or too many write tasks
+  ///   have been suspended.
+  internal func write(_ element: Element) async throws {
+    try await self.storage.write(element)
+  }
+
+  /// Write the final element
+  internal func finish(_ end: End) async throws {
+    try await self.storage.finish(end)
+  }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+fileprivate final class AsyncWriterStorage<
+  Element,
+  End,
+  Delegate: AsyncWriterDelegate
+> where Delegate.Element == Element, Delegate.End == End {
+  /// A value pending a write.
+  private struct Pending<Value> {
+    var value: Value
+    var continuation: UnsafeContinuation<Void, Error>
+  }
+
+  private typealias PendingElement = Pending<Element>
+  private typealias PendingEnd = Pending<End>
+
+  private enum CompletionState {
+    /// Finish hasn't been called yet. May move to `pending` or `completed`.
+    case incomplete
+    /// Finish has been called but the writer is paused. May move to `completed`.
+    case pending(PendingEnd)
+    /// The completion message has been sent to the delegate. This is a terminal state.
+    case completed
+
+    /// Move from `pending` to `completed` and return the `PendingCompletion`. Returns `nil` if
+    /// the state was not `pending`.
+    mutating func completeIfPending() -> PendingEnd? {
+      switch self {
+      case let .pending(pending):
+        self = .completed
+        return pending
+      case .incomplete, .completed:
+        return nil
+      }
+    }
+
+    var isPendingOrCompleted: Bool {
+      switch self {
+      case .incomplete:
+        return false
+      case .pending, .completed:
+        return true
+      }
+    }
+  }
+
+  /// A lock which must be help when accessing or modifying mutable state held by this storage.
+  private var lock: Lock
+
+  /// Elements and continuations which have been buffered but are awaiting consumption by the
+  /// delegate.
+  private var pendingElements: CircularBuffer<PendingElement>
+
+  /// The maximum number of pending elements. `pendingElements` must not grow beyond this limit.
+  private let maxPendingElements: Int
+
+  /// The completion state of the writer.
+  private var completionState: CompletionState
+
+  /// Whether the writer is paused.
+  private var isPaused: Bool = false
+
+  /// The delegate to process elements. By convention we call the delegate before resuming any
+  /// continuation.
+  private var delegate: Delegate
+
+  fileprivate init(maxPendingElements: Int, delegate: Delegate) {
+    self.lock = Lock()
+    self.maxPendingElements = maxPendingElements
+    self.pendingElements = CircularBuffer(initialCapacity: maxPendingElements)
+    self.completionState = .incomplete
+    self.delegate = delegate
+  }
+
```

Comment:
I think we need a `deinit` that validates final state. If the Storage is released and the user has not called `finish` is there some message we want to emit?

Reply:
That's a good point -- we _probably_ want an assertion/precondition.